### ISSUE TITLE
Value returned from malloc is not checked for NULL

### ISSUE THREAD

**Issue by Costava** [zakirullin/tiny-compiler#6]
In multiple files, `malloc` is called but the return value is not checked for NULL.
e.g.
https://github.com/zakirullin/tiny-compiler/blob/7b45c6a2f189822f5e56979e7a26bd689f5ecd92/src/stack.c#L7-L13

https://github.com/zakirullin/tiny-compiler/blob/7b45c6a2f189822f5e56979e7a26bd689f5ecd92/src/lex.c#L41-L47

Dereferencing a NULL pointer is undefined behavior and should be avoided.

The calls to `malloc` could be replaced with a wrapper function that checks the value returned from `malloc` and calls `exit` if there is an error.

**Comment by zakirullin** [comment]
That's true. Even though it's hardly reproducible on modern systems, it often happens on embed. So why give people wrong insight with malloc and undefined behaviors? You're right.
At first I thought it's unnecessary complication for the project like this, but OK, I've reconsidered that.

```c
#define SAFEMALLOC ...
if (head == NULL) {
  fprintf(stderr, "...");
  exit(EXIT_FAILURE);
} 
```

What do you think about macros like that? Or maybe you've come to better solution

**Comment by Costava** [comment]
A function would work perfectly fine in this case, so I would personally avoid using a macro.

C implementations are permitted to return `NULL` from `malloc(0)` so some of our options:
1. A comment above the function/macro mentioning not to pass 0 for the size argument and let the function/macro print an error and exit on systems that return `NULL` from `malloc(0)`
2. If the size argument is 0, the function/macro returns whatever it gets from `malloc` even if that is NULL
3. Explicitly print an error and exit if size is 0

I have been using option 1 but both 1 and 2 seem acceptable. I am not dead-set yet on which is preferable (I can imagine myself switching to using option 2).

**Comment by zakirullin** [comment]
Fixed - 64888dd89ceff98b4dfafc285a3631f40052d27d
